### PR TITLE
GH-325 - Use GitHub's GraphQL service

### PR DESCRIPTION
**Pull request by srgyrn**
#### Summary
This pull request adds functionality to use Github GraphQL API and replaces REST API with GraphQL API where pull request details are retrieved. 

The new graphql package communicates with GraphQL API via [https://github.com/shurcooL/githubv4](https://github.com/shurcooL/githubv4). Anything related to this third party package are encapsulated for decoupling reasons:
- Query struct needed by githubv4 can be created in Go code with the new graph/query.
- Response from githubv4 is converted to a local type Response (server/plugin/internal/graphql/response.go). All githubv4 types are converted to native data types.

#### Ticket Link
https://github.com/mattermost/mattermost-plugin-github/issues/325

### CODE REVIEW

**Comment by hanzei** [comment]
@srgyrn That you very much for raising this PR! Given the sheer size I will need a few days to properly review it.

**Comment by hanzei** [comment]
@reflog I'm wondering if you are interested in also reviewing this PR as your worked with GraphQL in the past IIRC.

**Comment by reflog** [comment]
@srgyrn thanks for your contribution!
Before we dive into the review - I wanted to ask - what prompted it?
What are we trying to fix? It seems (from a cursory glance at the code) that quite alot of new code was added, so I want to understand the benefit first.

Thanks again!

**Comment by larkox** [comment]
@reflog I was there when the ticket was created. The main problem is that for fetching all the needed information for the PRs and issues we now have to do 3 different calls for each open issue assigned. That can consume the easily the allowed rates and takes significant time to load.

Using GraphQL potentially will solve this, since we are able to get the whole information we need in just one query.

You can get a bit more of insight in the linked GH issue.

**Comment by reflog** [comment]
@larkox - awesome. Thanks for the details. I'll get on the review

**Comment by reflog** [comment]
@srgyrn question - what is the reason for implementing so many 'generic' graphql methods, when all we need to do is to construct a struct with proper tags, like:
```
var q struct {
	RepositoryOwner struct {
		Login        string
		Organization struct {
			Description string
		} `graphql:"... on Organization"`
		User struct {
			Bio string
		} `graphql:"... on User"`
	} `graphql:"repositoryOwner(login: \"github\")"`
}
```
and execute it?
I don't see the reason for having this 'query builder' pattern and scalar related stuff in this plugin.

**Comment by srgyrn** [comment]
Hey @reflog,

Having worked in projects before which communicate with Elasticsearch without a client like [Elastica](https://github.com/ruflin/Elastica), I've seen codebases end up getting polluted with so many, so long and sometimes redundant json queries. Also, using the structs expected by [githubv4](https://github.com/shurcooL/githubv4) within the project's code will make the project very dependent to this third party package. You can see in the [example structs given in githubv4's repo](https://github.com/shurcooL/githubv4#scalar-types), package specific data types like githubv4.String, githubv4.HTML or [enums](https://github.com/shurcooL/githubv4/blob/master/enum.go) are used. So anyone who wants to contribute to the project will have to know and/or learn about them too.

True, my suggested way makes query creating longer and involves more error checks than just writing down a struct. However, it leaves the project's dependency on a third party package just at the query builder bit. Creating scalar queries is all about copying and pasting names and values from the GraphQL Explorer. More complex queries (object, union...) is all about setting children in them.

With the query builder, you don't need to worry about package specific data types at all once required ones are added to the decorator/scalartypes.go. The same thing goes for struct tags as well. You can simply add a new one with validations so there won't be an unnecessary call to the GraphQL API if an unexpected value is set.

**Comment by reflog** [comment]
@srgyrn - in general, I agree with you. Having better abstractions is great for large projects.
This is however seems to me like an overkill for this plugin. But I am about 0/5 on this.
Let's pull a couple more people in for the discussion. cc @ethervoid @agnivade

**Comment by agnivade** [comment]
I really don't have any knowledge about GraphQL, but from a software design perspective - the tradeoff here is codebase size vs too many abstractions. For a big project that heavily uses GraphQL queries, it makes sense to write a low-level library to manage boilerplates in a proper way. But for a small project which uses GraphQL sparingly, I would lean towards using the structs to keep things simple and maintainable.

There is a cost to more code that we add. That has more chance of bugs, more lines of code to maintain and debug. The benefits of additional code should outweigh these costs.

I really don't have much context on the future direction that this plugin will take or how much more GraphQL might be used in future, so I will leave it to the plugin owners to take a call on that.

@srgyrn -
In any case, it is wonderful to see the effort that you have put in to write so much code and also tests for it. Thank you for your time :+1:

**Comment by ethervoid** [comment]
First of all, I love the work you've done here, and seeing the number of tests you made to cover it is amazing. Thank you very much

Regarding the PR itself, I agree with you in the base, it's really useful to decouple third-party libraries and I like the idea but I also agree with the comment from other reviewers, this looks a little overkill for this plugin. 

If I may suggest I'd do something like [this](https://github.com/mattermost/mattermost-mattermod/blob/master/server/github_client.go). Decouple the library using a service so we can mock that up and test it properly. You can also encapsulate the queries inside and in case the situation explodes as you said from past experiences we can connect a new "Client" and use the builder.

Again I want to stress that I think this is amazing work and has a lot of effort poured into it. Thank you

**Comment by srgyrn** [comment]
@reflog @agnivade @ethervoid Thank you for your time and feedback :)

@ethervoid Shall I remove the query builder and integrate the package like the example you've suggested or close the PR?

**Comment by reflog** [comment]
@srgyrn thank you!
Please don't close the PR, just simplify it if possible. @ethervoid suggestion sounds good

**Comment by ethervoid** [comment]
Agree with @reflog , you can work on this PR :) . Thank you!

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking on the discussion point above to get this out of my review queue.

Please ping me once the changes are in.

**Comment by jfrerich** [comment]
Can we get a consensus or proposal from the reviewers that will provide actionable items for the PR author.  

Thanks for your work, @srgyrn

**Comment by agnivade** [comment]
The suggested proposal is right here https://github.com/mattermost/mattermost-plugin-github/pull/404#issuecomment-732816812.

**Comment by srgyrn** [comment]
Hello everyone, I've been a bit busy at work but started working on the requested changes this weekend. I should be pushing them next week. I'm sorry I haven't been responsive enough.

**Comment by srgyrn** [comment]
Hello @hanzei, I did the requested changes. The golangci-lint fails but it seems to be a false positive: 

- https://github.com/golangci/golangci-lint/issues/1517
- https://gitlab.com/opennota/check#known-limitations

**Comment by jfrerich** [comment]
@hanzei, gentle ping for review :)

**Comment by jfrerich** [comment]
@srgyrn, some folks are just getting back from a break but will get to this shortly.  @hanzei, gentle reminder to review

File: server/plugin/internal/graphql/pullrequest.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		return res, err
+	}
+
+	var wg sync.WaitGroup
+
+	for _, resp := range query.Search.Nodes {
+		var requestedReviewers []string
+		var reviews []model.PullRequestReview
+
+		wg.Add(2)
```
Comment (hanzei) [inline server/plugin/internal/graphql/pullrequest.go:39]: Spawning two go routing to do a simple loop over an array seems excessive and will most likely not improve performance.

**Review by hanzei — CHANGES_REQUESTED** [review]
Great work @srgyrn :+1: 

I've leaved a couple of comments on the general code style. Maybe the PR can be simplified even more.

File: server/plugin/api.go [hunk 2]
```diff
@@ -609,104 +608,16 @@ func (p *Plugin) getPrsDetails(w http.ResponseWriter, r *http.Request, userID st
...
-	githubClient := p.githubConnect(*info.Token)
+	graphQLClient := p.graphQLConnect(info)
 
-	var prList []*PRDetails
-	if err := json.NewDecoder(r.Body).Decode(&prList); err != nil {
-		p.API.LogWarn("Error decoding PRDetails JSON body", "error", err.Error())
-		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: "Please provide a JSON object.", StatusCode: http.StatusBadRequest})
+	prDetailResponse, graphErr := graphQLClient.PullRequests.Get()
+	if graphErr != nil {
+		p.API.LogError("Failed to fetch PR details", "error", graphErr.Error())
```
[the remark this answers is not in the document]
Comment (srgyrn) [reply server/plugin/api.go:615]: I would but variable name _err_ is used in the same method and above this line and its type is _APIErrorResponse_ so there is a mismatch.

File: go.mod [hunk 3]
```diff
@@ -9,6 +9,9 @@ require (
 	github.com/mattermost/mattermost-plugin-api v0.0.12-0.20200908143138-66edf222f7ea
 	github.com/mattermost/mattermost-server/v5 v5.27.0
 	github.com/pkg/errors v0.9.1
+	github.com/shurcooL/githubv4 v0.0.0-20200928013246-d292edc3691b
+	github.com/shurcooL/graphql v0.0.0-20200928012149-18c5c3165e3a // indirect
```
[the remark this answers is not in the document]
Comment (srgyrn) [reply go.mod:13]: github.com/shurcooL/githubv4 uses that one and githubv4 doesn't have a go.mod file. Hence its dependencies are added to this repo's go.mod.

File: server/plugin/internal/graphql/client_test.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	}
+
+	err := client.executeQuery(&query, nil)
+	if err != nil {
+		t.Fatalf("executeQuery() err: %v", err)
+	}
+
+	if query.Viewer.Login == "" {
+		t.Errorf("struct field Login is empty, expected value")
+	}
```
[the remark this answers is not in the document]
Comment (srgyrn) [reply server/plugin/internal/graphql/client_test.go:40]: That's my bad. Fixing tests.

**Review by hanzei — APPROVED** [review]
Very nice work. LGTM :+1:

File: server/plugin/api.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/api.go:615]: Good point, missed that.

**Review by reflog — APPROVED** [review]
LGTM

**Comment by ethervoid** [comment]
@DHaussermann may you please take a look an finish the review cycle? Thx!

**Comment by mickmister** [comment]
@srgyrn Awesome work! This is a great performance improvement and simplifies the communication with GitHub. I have one comment on the functionality behavior.

Before this PR, the functionality worked like:

- The user visits a tab on the LHS such as "Your Open Pull Requests"
- The frontend sends a request to fetch PR details and provides a list of `PR number / repo` pairs (specific to the current tab's category)
- The backend performs an HTTP request per each pair

Now the backend is using GraphQL to fetch all at once, but it is not taking into account the pairs being sent by the client. Is the backend fetching/returning more than necessary in the GraphQL query? Or is this negligible due to the dataset generally being on the same order of size across each tab category? cc @hanzei

**Comment by DHaussermann** [comment]
@hanzei and @srgyrn I'm very sorry. I did not realize this was now only waiting on me. 
Can one of you please review the merge conflicts. I'll get this tested.

**Comment by larkox** [comment]
@srgyrn Would you mind resolving the conflicts? Thanks!

**Comment by hanzei** [comment]
I've pushed `master` to this branch, but while testing it a couple of issue were found:
1. The number of reviews is missing in the RHS
2. The PR CI information is missing.

**Pull request closed without merging**